File: Difficulty: Medium/ZigZag Tree Traversal/zig-zag-tree-traversal.cpp
/*
class Node {
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};
*/
class Solution {
  public:
  
    void solve(Node* root, int level, vector<vector<int>>& zvec) {
        if(root==nullptr) return;
        
        
        
        if(level&1) {
            if(level==zvec.size()) {
                zvec.push_back({root->data});
            }
            else {
                zvec[level].insert(zvec[level].begin(),root->data);
            }
        }
        else {
            if(level==zvec.size()) {
                zvec.push_back({root->data});
            }
            else {
                zvec[level].push_back(root->data);
            }
        }
        
        
        solve(root->left,level+1,zvec);
        solve(root->right,level+1,zvec);
        
    }
    vector<int> zigZagTraversal(Node* root) {
        // code here
        vector<vector<int>> zvec;
        vector<int> ans;
        
        
        int level = 0;
        
        solve(root,level,zvec);
        
        
        for(auto& v:zvec) {
            ans.insert(ans.end(),v.begin(),v.end());
        }
        
        return ans;
    }
};

File: Difficulty: Medium/Zig-Zag Tree Traversal/zig-zag-tree-traversal.java
/* Structure of Binary Tree Node
class Node {
    int data;
    Node left, right;
    Node(int d)
    {
        data = d;
        left = right = null;
    }
}*/

class Solution {
    
    private void solve(Node root, int level, ArrayList<ArrayList<Integer>> arr) {
       
       if(root==null) return;
       
        if(level == arr.size()) {
            arr.add(new ArrayList<Integer>());
            
        }
        
            arr.get(level).add(root.data);
        
        
        solve(root.left,level+1,arr);
        solve(root.right,level+1,arr);
        
    }
    ArrayList<Integer> zigZagTraversal(Node root) {
        // code here
        ArrayList<Integer> ans = new ArrayList<>();
        ArrayList<ArrayList<Integer>> ansi = new ArrayList<>();
        int level = 0;
        solve(root,0,ansi);
        
        int i = 0;
        for(ArrayList<Integer> arr: ansi) {
            if(i%2==1) {
                Collections.reverse(arr);
            }
            ans.addAll(arr);
            i++;
            
        }
        
        return ans;
        
    }
}